[RISCV] Enable early if-conversion

mgudim opened the pull request:
Implements the necessary target methods (insertSelect, canInsertSelect) and adds early if-conversion to the RISCV pipeline. Doing if-conversion increases the number of executed instructions, so it only makes sense if there is enough ILP.

mgudim committed:
[RISCV] Enable early if-conversion.

Implements the necessary target methods (insertSelect, canInsertSelect)
and adds early if-conversion to the RISCV pipeline. Doing if-conversion
increases the number of executed instructions, so it only makes sense if
there is enough ILP.

mgudim commented on llvm/lib/CodeGen/EarlyIfConversion.cpp:
```diff
@@ -879,7 +879,7 @@ bool EarlyIfConverter::shouldConvertIf() {
   // doesn't alias any of the memory-writes in the loop, but it is likely to
   // read to same value multiple times.
-  if (CurrentLoop && any_of(IfConv.Cond, [&](MachineOperand &MO) {
+  if (CurrentLoop && all_of(IfConv.Cond, [&](MachineOperand &MO) {
```
I created a separate PR for this change: https://github.com/llvm/llvm-project/pull/92405

topperc commented:
Tests?

topperc commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -1446,6 +1446,127 @@ RISCVInstrInfo::optimizeSelect(MachineInstr &MI,
+            NotCCReg)
+        .addReg(LHSReg)
+        .addReg(RHSReg);
+    BuildMI(MBB, MI, DL, get(RISCV::XORI), DstReg).addReg(NotCCReg).addImm(1);
```
Can we swap the true/false of the select to avoid the XORI?

mgudim replied:
Yes, thanks!

mgudim replied:
I did the swap in `insertSelect`.

mgudim committed:
Swap TrueReg and FalseReg to avoid negation in insertSelect

mgudim commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -1446,6 +1446,132 @@ RISCVInstrInfo::optimizeSelect(MachineInstr &MI,
+  TargetSchedModel SchedModel;
+  SchedModel.init(&STI);
+
+  CondCycles = getICmpCost(Cond[0].getImm(), SchedModel);
```
I am not sure I  understood correctly what `CondCycles` means. I need to double-check this.

mgudim committed:
Some bug fixes and added test.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVSubtarget.cpp:
```diff
@@ -203,3 +211,12 @@ unsigned RISCVSubtarget::getMinimumJumpTableEntries() const {
+  SchedModel.init(this);
+  return RISCVForceEalyIfcvt ||
+         (!RISCVDisableEarlyIfcvt &&
+          (hasStdExtZicond() || hasVendorXVentanaCondOps()) &&
```
TODO: Add support for `xtheadcondmov`? cc @zixuan-wu

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -1446,6 +1446,196 @@ RISCVInstrInfo::optimizeSelect(MachineInstr &MI,
+    CC = (CC == RISCVCC::COND_GE) ? RISCVCC::COND_LT : RISCVCC::COND_LTU;
+    std::swap(CondZeroEqzReg, CondZeroNezReg);
+  }
+  insertICmp(MBB, MI, DL, CCReg, CC, LHSReg, RHSReg);
```
We may not need this for some cases.

mgudim replied:
> We can remove the xor+seqz here.

Done

wangpc-pp reviewed (commented):
I just have evaluated this patch, and I think there is a room for more optimizations:
```c
int loop(int m, int n, int *randArray) {
  int t = 0;
  for (int j = 0; j < n; j++) {
    for (int i = 0; i < 4 + m; i++) {
      if (randArray[i]) {
        t += 3 + 3 * t;
      } else {
        t -= 1 - 5 * t;
      }
    }
  }
  return t;
}
```
For this case, I compared the results of two configurations:
1.  `clang -O2 -march=rv64gc_zba_zbb_zbc_zbs_zicond -mllvm -two-entry-phi-node-folding-threshold=16`. I increased the threshold so that we can do if conversion in LLVM IR level via `SimplifyCFG`.
2.  `clang -O2 -march=rv64gc_zba_zbb_zbc_zbs_zicond -mllvm -riscv-disable-early-ifcvt=false -mllvm -riscv-force-early-ifcvt=true -mllvm -stress-early-ifcvt=true`.
The diff is:
![图片](https://github.com/llvm/llvm-project/assets/137158460/177da652-3b3d-4c02-86b5-0575960aa75c)

We can remove the `xor`+`seqz` here.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -1446,6 +1446,196 @@ RISCVInstrInfo::optimizeSelect(MachineInstr &MI,
+
+  Register TrueValOrZeroReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+  Register FalseValOrZeroReg = MRI.createVirtualRegister(&RISCV::GPRRegClass);
+  BuildMI(MBB, MI, DL, get(CondZeroEqzOpc), TrueValOrZeroReg)
```
We may need to implement all the patterns in SDAG ISel here to reduce some instructions.

wangpc-pp replied:
Or, we implement these ISel patterns in a MIR pass, and then run the pass after early if-conversion pass.

mgudim replied:
These patterns apply when one of the sides is zero. In our case we are looking at select between the results of two basic blocks. How often will it be that the result of a basic block is just zero?

wangpc-pp replied:
I have no data to show how common these cases will be, but we have do a lot of peepholes in `RISCVInstrInfoZicond.td` and `RISCVTargetLowering::lowerSELECT` to reduce the comparison and Zicond instructions.
After if-conversion, we need to do the same peepholes (which means we may not need czero.eqz+czero.nez+or to simulate `select`).

mgudim replied:
I see what you mean. This seems like a lot of work and code duplication to reimplement all of these here.  Here's a list of all peepholes that we do in `lowerSELECT`:

```
(select c, t, 0) -> (czero_eqz t, c)
(select c, 0, f) -> (czero_nez f, c)
(select c, (and f, x), f) -> (or (and f, x), (czero_nez f, c))
(select c, t, (and t, x)) -> (or (czero_eqz t, c), (and t, x))
(select c, c1, c2) -> (add (czero_nez c2 - c1, c), c1)
(select c, c1, c2) -> (add (czero_eqz c1 - c2, c), c2)

(select c, -1, y) -> -c | y
(select c, y, -1) -> (c-1) | y
(select c, 0, y) -> (c-1) & y
(select c, y, 0) -> -c & y
select c, ~x, x --> xor -c, x
(select x, x, y) -> x | y
(select !x, x, y) -> x & y
(select x, y, x) -> x & y
(select !x, y, x) -> x | y
binOp ((select cond, x, c0), c1) -> select cond binOp(x, c1), binOp(c0, c1)
```
But perhaps we can address this later in a different MR, this MR is just the basic functionality.

wangpc-pp replied:
> This seems like a lot of work and code duplication to reimplement all of these here. 

Yes! So I said `Or, we implement these ISel patterns in a MIR pass and then run the pass after early if-conversion pass`. :-)

> But perhaps we can address this later in a different MR, this MR is just the basic functionality.

Agree. Can you make this PR non-draft? I ran llvm-testsuite and some internel tests and I haven't found any correctness issue, so I think it's ready for reviewing.

mgudim replied:
This needs to be merged first: https://github.com/llvm/llvm-project/pull/95877
This MR uses a hack to bypass that.

Also I need to add some tests too.

wangpc-pp replied:
It's OK to be stacked on other PRs, make it non-draft so that more reviewers can see it.

mgudim commented:
@wangpc-pp Thanks for the test case, I'll look into that. I have only tested on x264 train workload and I got 1% impovement in cycle count. Can you share your results?

mgudim commented:
This is a  fix for detecting when condition is loop-invariant: https://github.com/llvm/llvm-project/pull/95877

That needs to be merged before this patch.

wangpc-pp commented:
> @wangpc-pp Thanks for the test case, I'll look into that. I have only tested on x264 train workload and I got 1% impovement in cycle count. Can you share your results?

I haven't run any particular benchmark and I just evaluated it on some internel microbenches. We can see some cycles reduction.

mgudim committed:
Simplified insertICmp for the cases of seqz and snez.
Updated cost estimate for these cases.
Made `getICmpCost` and `insertICmp` accept `Cond` to make their
signatures more like `insertSelect`.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVInstrInfo.cpp:
```diff
@@ -1446,6 +1446,199 @@ RISCVInstrInfo::optimizeSelect(MachineInstr &MI,
+  case RISCVCC::COND_LTU:
+    return SchedModel.computeInstrLatency(RISCV::SLTU);
+  case RISCVCC::COND_EQ:
+    return SchedModel.computeInstrLatency(RISCV::XOR) +
```
The cost is inaccurate now as we may not emit a XOR.

wangpc-pp commented on llvm/lib/Target/RISCV/RISCVSubtarget.cpp:
```diff
@@ -203,3 +211,12 @@ unsigned RISCVSubtarget::getMinimumJumpTableEntries() const {
+  return RISCVForceEalyIfcvt ||
+         (!RISCVDisableEarlyIfcvt &&
+          (hasStdExtZicond() || hasVendorXVentanaCondOps()) &&
+          SchedModel.hasInstrSchedModelOrItineraries());
```
Do we really need `hasInstrSchedModelOrItineraries` here? We have an option to force early if-conversion, the schedule model is not really needed here.

mgudim committed:
Updated tests.
